trin-core/src/utp/packets.rs

```diff
@@ -0,0 +1,322 @@
+    connection_id: u16,
+    // This time is in microseconds
+    timestamp: u32,
+    timestamp_difference: u32,
+    wnd_size: u32,
+    pub seq_nr: u16,
+    ack_nr: u16,
+}
+
+impl PacketHeader {
+    pub fn new() -> Self {
+        PacketHeader {
+            type_ver: u8::from(PacketType::Data) << 4 | VERSION,
+            extension: 0,
+            connection_id: 0,
+            timestamp: 0,
+            timestamp_difference: 0,
+            wnd_size: 0xf000,
+            seq_nr: 0,
+            ack_nr: 0,
+        }
+    }
+
+    pub fn encode(&mut self) -> Vec<u8> {
+        let mut buf = vec![];
+        buf.extend_from_slice(&[self.type_ver]);
+        buf.extend_from_slice(&[self.extension]);
+        buf.extend_from_slice(&self.connection_id.to_be_bytes());
+        buf.extend_from_slice(&self.timestamp.to_be_bytes());
+        buf.extend_from_slice(&self.timestamp_difference.to_be_bytes());
+        buf.extend_from_slice(&self.wnd_size.to_be_bytes());
+        buf.extend_from_slice(&self.seq_nr.to_be_bytes());
+        buf.extend_from_slice(&self.ack_nr.to_be_bytes());
+        buf
+    }
+
+    pub fn decode(bytes: &[u8]) -> PacketHeader {
+        PacketHeader {
+            type_ver: bytes[0],
+            extension: bytes[1],
+            connection_id: u16::from_be_bytes([bytes[2], bytes[3]]),
+            timestamp: u32::from_be_bytes([bytes[4], bytes[5], bytes[6], bytes[7]]),
+            timestamp_difference: u32::from_be_bytes([bytes[8], bytes[9], bytes[10], bytes[11]]),
+            wnd_size: u32::from_be_bytes([bytes[12], bytes[13], bytes[14], bytes[15]]),
+            seq_nr: u16::from_be_bytes([bytes[16], bytes[17]]),
+            ack_nr: u16::from_be_bytes([bytes[18], bytes[19]]),
+        }
+    }
+
+    pub fn version(&self) -> u8 {
+        self.type_ver & 0xf
+    }
+    pub fn set_version(&mut self, int: u8) {
+        self.type_ver = (self.type_ver & 0xf0) | (int & 0xf)
+    }
+    pub fn type_(&self) -> PacketType {
+        PacketType::try_from(self.type_ver >> 4).unwrap()
+    }
+    pub fn set_type(&mut self, t: PacketType) {
+        self.type_ver = (self.type_ver & 0xf) | (u8::from(t) << 4)
+    }
+
+    pub fn set_connection_id(&mut self, connection_id: u16) {
+        self.connection_id = connection_id;
+    }
+
+    pub fn set_timestamp(&mut self, timestamp: u32) {
+        self.timestamp = timestamp;
+    }
+
+    pub fn set_timestamp_difference(&mut self, timestamp_difference: u32) {
+        self.timestamp_difference = timestamp_difference;
+    }
+
+    pub fn set_wnd_size(&mut self, wnd_size: u32) {
+        self.wnd_size = wnd_size;
+    }
+
+    pub fn set_seq_nr(&mut self, seq_nr: u16) {
+        self.seq_nr = seq_nr;
+    }
+
+    pub fn set_ack_nr(&mut self, ack_nr: u16) {
+        self.ack_nr = ack_nr;
+    }
+}
+
+#[derive(Debug)]
+pub struct Extension {
+    pub extension_type: u8,
+    pub len: u8,
+    pub bitmask: Vec<u8>,
+}
+
+#[derive(Clone, Debug)]
+pub struct Packet(pub Vec<u8>);
```

Comment:
To get rid of the `get_header()` repetitions, we can add `header` and a `payload` fields in this struct and then update those on initialization with `new` or `with_payload` methods.

Reply:
Because of how data is handled after the header, it would just be cleaner to store the header with the vector as a tuple then anything that would be stored after the header would be in the Vec extension payload etc.

I like this solution because it is very readable though